File: src/ai/RAG/NestJS/prompt-builder.ts
import { readFileSync } from 'fs';
import { RetrievalRecipe } from './retrieval.service';

export class PromptBuilder {
  constructor(private readonly template: string) {
    if (!template?.trim()) {
      throw new Error('Prompt template is required');
    }
  }

  static fromFile(templatePath: string): PromptBuilder {
    return new PromptBuilder(readFileSync(templatePath, 'utf8'));
  }

  build(question: string, recipes: RetrievalRecipe[]): string {
    return this.template
      .replace('{recipes}', this.formatRecipes(recipes))
      .replace('{question}', question.trim());
  }

  private formatRecipes(recipes: RetrievalRecipe[]): string {
    if (recipes.length === 0) {
      return 'No relevant recipes were retrieved.';
    }

    return recipes.map((recipe) => this.formatRecipe(recipe)).join('\n\n');
  }

  private formatRecipe(recipe: RetrievalRecipe): string {
    return [
      `Recipe ID: ${recipe.recipeId}`,
      `Title: ${recipe.title}`,
      `Description: ${recipe.description || 'Not specified'}`,
      `Tags: ${recipe.tags || 'Not specified'}`,
      `Ingredients: ${recipe.ingredients.length > 0 ? recipe.ingredients.join(', ') : 'Not specified'}`,
      `Preparation steps: ${recipe.preparationSteps || 'Not specified'}`,
      `Cooking time: ${recipe.cookingTime || 'Not specified'}`,
      `Similarity score: ${recipe.similarityScore.toFixed(6)}`,
    ].join('\n');
  }
}

File: src/ai/RAG/Python/prompt_builder.py
from __future__ import annotations

from pathlib import Path


DEFAULT_TEMPLATE_PATH = Path(__file__).resolve().parents[1] / "prompts" / "rag_prompt.txt"


class PromptBuilder:
    def __init__(self, template_path: Path | None = None):
        self.template_path = template_path or DEFAULT_TEMPLATE_PATH

    def build(self, question: str, recipes: list) -> str:
        template = self.template_path.read_text(encoding="utf-8")
        return template.replace("{recipes}", self._format_recipes(recipes)).replace("{question}", question.strip())

    def _format_recipes(self, recipes: list) -> str:
        if not recipes:
            return "No relevant recipes were retrieved."

        return "\n\n".join(self._format_recipe(recipe) for recipe in recipes)

    def _format_recipe(self, recipe) -> str:
        ingredients = ", ".join(recipe.ingredients) if recipe.ingredients else "Not specified"
        return "\n".join([
            f"Recipe ID: {recipe.recipeId}",
            f"Title: {recipe.title}",
            f"Description: {recipe.description or 'Not specified'}",
            f"Tags: {recipe.tags or 'Not specified'}",
            f"Ingredients: {ingredients}",
            f"Preparation steps: {recipe.preparationSteps or 'Not specified'}",
            f"Cooking time: {recipe.cookingTime or 'Not specified'}",
            f"Similarity score: {recipe.similarityScore:.6f}",
        ])